crates/bevy_asset/src/asset_server.rs

```diff
@@ -462,22 +480,20 @@ impl AssetServer {
                                     source_info.load_state = LoadState::Loaded;
                                 }
                             }
+                        } else {
+                            warn!("`AssetSource` is not found");
                         }
                     }
 
                     let _ = assets.set(result.id, result.asset);
                 }
                 Ok(AssetLifecycleEvent::Free(handle_id)) => {
-                    if let HandleId::AssetPathId(id) = handle_id {
+                    if let HandleId::AssetPathId(id) = &handle_id {
                         let asset_sources = asset_sources_guard
                             .get_or_insert_with(|| self.server.asset_sources.write());
-                        if let Some(source_info) = asset_sources.get_mut(&id.source_path_id()) {
-                            source_info.committed_assets.remove(&id.label_id());
-                            if source_info.is_loaded() {
-                                source_info.load_state = LoadState::Loaded;
-                            }
-                        }
+                        asset_sources.remove(&id.source_path_id());
```

Comment:
I have a feeling this line is incorrect due to the multi-threaded nature of Bevy. 
If you look inside `load_async`, the `SourceInfo` struct get created, and then after that the function **assumes** the `SourceInfo` exists for the rest of the function. 

I'm pretty sure it's possible for the final handle to get dropped, but before `free_unused_assets` is called, another thread calls `load_async` which grabs the current `SourceInfo`. 
`free_unused_assets` is called, which deletes the `SourceInfo` concurrently, and then the `load_async` `unwraps()` the `SourceInfo` causing a crash.

If anyone else has knowledge on this, I would love to hear it :smile:

Reply:
Edit: 
I recently implemented a `LoadState::Removing` status, to ensure the `SourceInfo` and other data only get removed if no other thread is trying to load the same asset.